test/crd.go

```diff
@@ -260,3 +263,34 @@ func ManualService(svc *v1alpha1.Service) *v1alpha1.Service {
 // environment without conflicting. This method will seed rand with the current time when
 // called for the first time.
 var AppendRandomString = helpers.AppendRandomString
+
+// ServiceNameForTest generates a random service name based on the test name.
+func ServiceNameForTest(t *testing.T) string {
+	return AppendRandomString(makeK8sNamePrefix(strings.TrimPrefix(t.Name(), "Test")))
+}
+
+// SubServiceNameForTest generates a random service name based on the test name and
+// the given subservice name.
+func SubServiceNameForTest(t *testing.T, subsvc string) string {
+	fullPrefix := strings.TrimPrefix(t.Name(), "Test") + "-" + subsvc
+	return AppendRandomString(makeK8sNamePrefix(fullPrefix))
+}
+
+// makeK8sNamePrefix converts each chunk of non-alphanumeric character into a single dash
+// and also convert camelcase tokens into dash-delimited lowercase tokens.
+func makeK8sNamePrefix(s string) string {
+	var sb strings.Builder
+	newToken := false
+	for idx, c := range s {
+		if !(unicode.IsLetter(c) || unicode.IsNumber(c)) {
+			newToken = true
+			continue
+		}
+		if idx > 0 && sb.Len() > 0 && (newToken == true || unicode.IsUpper(c)) {
```

Comment:
```suggestion
		if sb.Len() > 0 && (newToken == true || unicode.IsUpper(c)) {
```
`sb.Len()` cannot be non-0, if index is 0.

Reply:
If the beginning has non-alphanumeric characters that may happen.

Reply:
I don't see how index can be 0, while sb isn't? :-) since if idx=0 then no characters have been added yet (first addition will be below). So the only time we can add a "-" if there is a valid k8s prefix already, which can happen only if index>=1

Reply:
That's true, I was thinking of the reverse (`sb.Len() = 0` while `idx > 0`)

Reply:
So, we can accept my suggestion, I guess? :)